src/stdx.zig

```diff
@@ -237,6 +237,55 @@ pub const log = if (builtin.is_test)
 else
     std.log;
 
+/// Compare two values by directly comparing the underlying memory.
+///
+/// Assert at compile time that this is a reasonable thing to do for a given `T`. That is, check
+/// that:
+///   - `T` doesn't have any non-deterministic padding.
+///   - `T` doesn't embed any pointers.
+pub fn eql_repr(comptime T: type, a: *const T, b: *const T) bool {
+    comptime assert(std.meta.trait.hasUniqueRepresentation(T));
+    comptime assert(pointer_free(T));
+    comptime assert(@sizeOf(T) * 8 == @bitSizeOf(T));
+
+    // Pick the biggest "word" for word-wise comparison, and don't try to early-return on the first
+    // mismatch, so that compiler can vectorize the loop.
+
+    const Int = inline for (.{ u64, u32, u16, u8 }) |Int| {
+        if (@alignOf(T) >= @alignOf(Int) and @sizeOf(T) % @sizeOf(Int) == 0) break Int;
+    } else unreachable;
+
+    const a_words = std.mem.bytesAsSlice(Int, std.mem.asBytes(a));
+    const b_words = std.mem.bytesAsSlice(Int, std.mem.asBytes(b));
+    assert(a_words.len == b_words.len);
+
+    var total: Int = 0;
+    for (a_words) |a_word, i| {
+        const b_word = b_words[i];
+        total |= a_word ^ b_word;
+    }
+
+    return total == 0;
+}
+
+fn pointer_free(comptime T: type) bool {
```

Comment:
Nit, but when I hear "free" in the context of pointers, I typically think of alloc/free.
But in this case it is pretty easy to tell what is going on, so it's ok as-is.